test/cases/testdata/functions/test-functions-default.yaml

```diff
@@ -1,21 +1,79 @@
 cases:
 - data:
   modules:
-  - |
-    package p.m
-
-    default hello = false
-
-    hello() = m {
-      m = input.message
-      1 == 2
-      m = "world"
-    }
-    h = m {
-      m = hello()
-    }
-  note: functions/default # not supported but shouldn't panic
-  query: data.p.m = x
+    - |
+      package test
+
+      default f(x) = 1
```

Comment:
Not sure I understand the policy change. Lets say you have a policy like:

```rego
package test

default f(1) = 1

f(x) = x {
   x > 0
}
```

```
opa eval -d policy.rego 'data.test.f(1)' -f pretty
1
```

```
opa eval -d policy.rego 'data.test.f(-1)' -f pretty
undefined
```

Would you have expected something else?

Reply:
I was just wondering about the semantics of default functions and their arguments. So, it seems like
> When every function body fails to evaluate to something, the default function is evaluated. The same semantics as for other function bodies applies: arguments need to match.

Just wondering because that's the main difference compared to complete rules -- we have arguments to deal with.

I guess you would have something like
```rego
default f(_) = "whatever"
```
most of the times. 🤔 

> ❓ May this value depend on x?

The would be the other side where arguments matter, rhs of the definition. For complete rules, I think some values are forbidden? I.e. what about

```rego
default f(x) = x > 1

f(x) = 1 if false
```
If `x` was allowed on the RHS, then "default functions" behave quite a bit differently than "default rules" -- their values aren't guaranteed as strongly, we'd only guarantee that they're evaluated.

Reply:
Yeah, default rules can’t depend on external refs like input either. Perhaps we should have default functions follow that principle too?

Reply:
hmm but thinking about it some more, my example is bogus:
```rego
default f(x) = x > 1
```
is completely OK, it just happens to evaluate to false. But it always _has_ a value.

Where this **could go wrong** is in cases like
```rego
default f(x) = x.foo
```